_posts/2020-11-11--kinesis-streams-part-2.md

```diff
@@ -0,0 +1,222 @@
+
+If u are inclined to do so, you can use an [EMR cluster as your custom consumer](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-kinesis.html). Or, as with writing to the stream, there are a lot of third-party tools that integrate with _Kinesis Streams_. Think of possibilities like Databricks and Apache Spark integration (which actually uses KCL under the hood), among others.
+
+On a perhaps more familiar side of things, you can have your own consumer application reading from the stream. For that purposes you can once again use either AWS SDK or _Kinesis Consumer Library (KCL)_, which is the counterpart of the _[already familiar KPL](https://dev.solita.fi/2020/05/28/kinesis-streams-part-1.html)_ and which I’ll be mentioning briefly.
+
+And finally, you can use AWS Lambda as you stream consumer, which gives you all the scalability and resilience that comes with it. You might have already guessed that that’s exactly what I’ll be focusing on in this post.
+
+But first, let’s lay out some ground for it.
+
+
+## Consuming the stream, two ways
+
+As with writing to a stream, when reading from it you should be aware that Kinesis stream is composed of shards (link to part1) and each shard in your stream comes with a limited throughput: you can write **1 MB or 1 000 records** of data per second, while on the way out, each shard provides up to **2 MB** of data per second. So, in theory, you can consume the stream **twice** as fast as you can write to it.
+
+
+### Shared throughput
+
+But not only that. This also means that you can actually have **at least two different consumers** reading from a same Kinesis Stream at all times. A thing to note here is that by default, the 2 MBs are **shared** between all the consumers you attach to the stream. So, if you would want to have more than recommended two to three consumers, you may face a situation when the data is being written to the stream faster than it can be read. This, in turn, may lead to some of the data expiring (aka being deleted from the stream) before any of the consumers get a chance to read it. You can think of it as a sink that u pour water into: if u do it faster than it can be drained, it ends up overflowing and u get water all over your floor. 
+
+**TODO**: image with two lambdas
+
+**TODO**: image of an overflowing bathtub? 🙂 
+
+To add more gravity to the situation, both, direct integration with _Kinesis Data Analytics_ and _Kinesis Firehose_, use this shared throughput of 2MB when acting as stream consumers. So you are supposed to add just one custom consumer to the stream if you are using any of those direct integrations.
+
+But wait: what if you r feeling greedy and want even more ways to consume your data from a single stream? Fex. you may have a firehose writing data to S3, a _Kinesis Analytics_ application doing some real-time data aggregations and then you also have a lambda function for custom processing.
```

Comment:
Took me a few seconds to figure out Fex meant for example. Maybe it's commonly used, but haven't seen before